File: server/game.js
/** game dimensions are 1000, 1000, 1000
/** game dimensions are 1000, 1000, 1000
 * @type {{start: module.exports.start, addPlayer: module.exports.addPlayer}}
 */

const config = {
    dimensions: [1000, 1000, 1000],
    startRadius: 30,
    food: {
        num: 500,
        radius: 10
    },
    viruses: {
        num: 20,
        radius: 50,
        colour: '#00ff00'
    },
    updatesPerSecond: 60
}

const types = {
    food: 'f',
    virus: 'v',
    player: 'p',
    bot: 'b'
}

const things = {}
let snapshot = {}

let nextId = 0

function massFromRadius(radius) {
    return (4 / 3) * Math.PI * (Math.pow(radius, 3))
}

function radiusFromMass(mass) {
    return Math.cbrt(mass / ((4 / 3) * Math.PI))
}

function euclideanDistance(t1, t2) {
    return Math.sqrt(Math.pow((t2.x - t1.x), 2) + Math.pow((t2.y - t1.y), 2) + Math.pow((t2.z - t1.z), 2))
}

function contains(thing1, thing2) {
    const d = euclideanDistance(thing1, thing2)
    return (d + thing2.r) < thing1.r
}

function getThingsOfType(things, type) {
    return Object.keys(things)
        .filter(k => things[k].t === type)
        .map(k => things[k])
}

function checkCollisions(things) {
    const players = getThingsOfType(things, types.player),
        food = getThingsOfType(things, types.food),
        both = players.concat(food)

    players.forEach(p1 => {
        both.forEach(b1 => {
            if (p1 !== b1 && contains(p1, b1)) {
                delete things[b1.id]
                const mass = p1.m + b1.m,
                    rad = radiusFromMass(mass)
                things[p1.id] = Object.assign(p1, {
                    m: mass,
                    r: rad
                })
                console.log(`p1 changed to ${JSON.stringify(things[p1.id])}`)
                console.log('nom nom nom')
            }
        })
    })
    return things
}

function randomPosition(){
    const [x, y, z] = config.dimensions
    return {
        x: ( Math.random() - 0.5 ) * x,
        y: ( Math.random() - 0.5 ) * y,
        z: ( Math.random() - 0.5 ) * z
    }
}

function middle() {
    return {
        x: 250,
        y: 250,
        z: 250
    }
}

function addPlayer(player) {
    let p = things[player.id]
    if (!p) {
        p = Object.assign({
            c: '#ff0000',
            r: config.startRadius,
            t: types.player,
            m: massFromRadius(config.startRadius)
        }, player, randomPosition())
        things[player.id] = p
    }
    return p
}

function randomColour(){
    return '#' + (Math.random() * 0xFFFFFF << 0).toString(16)
}

function repeatedly(n, fn){
    for (let i = 0; i < n; i++){
        fn()
    }
}

function snapshotState(state) {
    const c = {}
    Object.keys(state).forEach(k => c[k] = state[k])
    return c
}

function diff(prev, next) {
    return {
        deleted: Object.keys(prev).filter(k => next[k] === undefined),
        changed: Object.keys(next)
            .filter(k => prev[k] === undefined || prev[k] !== next[k])
            .map(k => next[k])
    }
}

function delta() {
    const d = diff(snapshot, checkCollisions(things))
    snapshot = snapshotState(things)
    return d
}

module.exports = {
    start: (updateFn) => {
        console.log('start game')
        repeatedly(config.food.num, () => {
            const f = Object.assign({
                c: randomColour(),
                id: ++nextId,
                t: types.food,
                r: config.food.radius,
                m: massFromRadius(config.food.radius)
            }, randomPosition())
            things[f.id] = f
        })
        repeatedly(config.viruses.num, () => {
            const v = Object.assign({
                c: config.viruses.colour,
                id: ++nextId,
                t: types.virus,
                r: config.viruses.radius,
                m: massFromRadius(config.viruses.radius)
            }, randomPosition())
            things[v.id] = v
        })
        setInterval(updateFn, 1000 / config.updatesPerSecond)
        return delta()
    },
    updatePosition: (id, pos) => {
        if (things[id]) {
            things[id] = Object.assign({}, things[id], pos)
        }
    },
    delta: delta,
    removePlayer: id => {
        delete things[id]
    },
    addPlayer: player => {
        const p = addPlayer(player)
        return {
            me: p,
            things: diff({}, things).changed,
            config: {
                foodRadius: config.food.radius,
                virusRadius: config.viruses.radius,
                virusColour: config.viruses.colour
            }
        }
    }
}

File: src/agario3d/game.clj
(ns agario3d.game
  (:require [clojure.core.async :refer [>! <! alts! chan close! go go-loop timeout]]
            [schema.core :as s]
            [agario3d.config :refer [config]]
            [clojure.math.numeric-tower :refer [expt]]
            [agario3d.loop :refer [every]]))

(def Pos
  "Scheme for position vector"
  {:x s/Num :y s/Num :z s/Num})

(def Agent
  "Schema for an agent in the game"
  {:c s/Str
   (s/optional-key :n) s/Str
   :id s/Any  ;;this could be a string or a num
   :t s/Keyword
   :r s/Num
   :m s/Num
   :x s/Num
   :y s/Num
   :z s/Num})

(def next-id (atom 0))

(def pi (. Math PI))

(s/defn euclidean-distance :- s/Num
  [p1 :- Pos p2 :- Pos]
  (let [dx (expt (- (:x p2) (:x p1)) 2)
        dy (expt (- (:y p2) (:y p1)) 2)
        dz (expt (- (:z p2) (:z p1)) 2)]
    (-> (+ dx dy dz)
        (expt ,, (/ 1 2)))))

(s/defn contains? :- s/Bool 
  [a1 :- Agent a2 :- Agent]
  (let [d (euclidean-distance a1 a2)]
    (-> (+ d (:r a2))
        (< ,, (:r a1)))))

(s/defn mass->radius :- s/Num
  [mass :- s/Num]
  (-> mass
      (/ ,, (* pi (/ 4 3)))
      (expt ,, (/ 1 3))))

(s/defn radius->mass :- s/Num 
  [radius :- s/Num]
  (-> radius
      (expt ,,, 3)
      (* ,, pi)
      (* ,, (/ 4 3))))

(s/defn random-coord :- s/Num
  [limit]
  (-> (rand)
      (- ,, 0.5)
      (* ,, limit)))

(s/defn random-pos :- Pos
  []
  (let [[x y z] (:dimensions config)]
    {:x (random-coord x)
     :y (random-coord y)
     :z (random-coord z)}))

(s/defn create-food  :- Agent []
  (let [radius (get-in config [:food :radius])]
    (merge  {:c "0xffffff"
             :id (swap! next-id inc)
             :t :food
             :r radius
             :m (radius->mass radius)} (random-pos))))

(s/defn create-virus  :- Agent []
  (let [radius (get-in config [:viruses :radius])]
    (merge  {:c (get-in config [:viruses :colour])             
             :id (swap! next-id inc)
             :t :virus
             :r radius
             :m (radius->mass radius)} (random-pos))))

(s/defn create-bot  :- Agent []
  (let [radius (:startRadius config)]
    (merge  {:c (get-in config [:bots :colour])
             :id (swap! next-id inc)
             :t :bot
             :r radius
             :m (radius->mass radius)} (random-pos))))

(s/defn create-player  :- Agent [player]
  (let [radius (:startRadius config)]
    (merge  {:c (:colour player)
             :n (:name player)
             :id (:id player)
             :t :player
             :r radius
             :m (radius->mass radius)} (random-pos))))

(s/defn create-agents :- [Agent] [n agentFn]
  (map (fn [n] (agentFn)) (range n)))

(s/defn things-of-type :- [Agent]
  [game :- s/Any t :- s/Keyword]
  (filter (fn [[k v]] (= t (:t v))) game))

(defn eat [eater eatee]
  (let [mass (+ (:m eater) (:m eatee))]
    (assoc eater :m mass :r (mass->radius mass))))

(defn get-collisions [game]
  (let [players (things-of-type game :player)
        bots (things-of-type game :bot)
        agents (concat players bots)
        food (things-of-type game :food)
        both (concat agents food)]
    (for [a agents
          b both
          :when (and (not= a b) (contains? a b))]
      [a b])))

(defn perform-collisions [game]
  (let [collisions (get-collisions game)]
    (reduce 
      (fn [g [eater eatee]]
        (prn "nom nom nom")
        (-> g
            (dissoc ,, (:id eatee))
            (assoc ,, (:id eater) (eat eater eatee)))) game collisions)))

(defn create-new-game
  "Seed the game with food, viruses and bots"
  []
  (prn "creating a brand new game")
  (atom  (let [foodNum (get-in config [:food :num])
               botNum (get-in config [:bots :num])
               virusNum (get-in config [:viruses :num])
               agents (concat (create-agents foodNum create-food)
                              (create-agents botNum create-bot)
                              (create-agents virusNum create-virus))]
           (reduce (fn [g a] (assoc g (:id a) a)) {:snapshot {}} agents))))

(defn create-snapshot [game]
  (->> (reduce (fn [snap [k v]]
                 (assoc snap k v)) {} game)
        (assoc game :snapshot ,,,)))

(defn create-diff [s game]
    (reduce (fn [diff [k v]]
                   (let [snap-val (get k s)]
                     (if (and (not= :snapshot k) 
                              (or (nil? snap-val) (not= snap-val v)))
                       (assoc diff k v)
                       diff))
                   ) {} game))

(defn move-bots [game])

(defn move-food [game])

(defn replenish-bots [game])

(defn replenish-food [game])

(defn update-game [game delta]
  (let [snap (:snapshot @game)]
    (->> (swap! game
                (fn [g]
                  (-> g
                      move-bots
                      move-food
                      replenish-bots
                      replenish-food
                      perform-collisions
                      create-snapshot)))
         (create-diff snap ,,,))))

(defn start-game [game]
  (go
    (let [tick (every (/ 1000 (:updatesPerSecond config)))]
      (loop [game game]
        (let [delta (<! tick)]
          (recur (update-game game delta)))))))

(defn update-player-position [game command]
  (if-let [player (get @game (:id command)) ]
    (let [updated (merge player command)]
      (swap! game assoc (:id command) updated)
      game)
    game))

(defn player-command [game command]
  (case (:type command)
    :position (update-player-position game command)
    game))

(defn add-player [game player]
  (let [p (create-player player)]
    (swap! game assoc (:id p) p)
    game))

(defn remove-player [game {id :id}]
  (swap! game dissoc id)
  game)

